internal/dashboard/business/ac/account.go

```diff
@@ -81,15 +82,60 @@ func ValidateAccount(ctx context.Context, username, password string) (*acmodel.A
 		account.Roles = roles
 	}
 
-	now := time.Now().Unix()
-	err = updateUserCredentials(ctx, credentials, username, account.password, account.Nickname, strconv.FormatInt(now, 10), account.Description)
-	if err != nil {
-		logrus.WithError(err).Warn("failed to update user credentials")
+	if account.LastLoginAt != nil && !account.LastLoginAt.IsZero() {
+		enforcer.accMu.Lock()
+		defer enforcer.accMu.Unlock()
+		// Update latest login time
+		now := time.Now().Unix()
+		up := &acmodel.UpdateAccountCreds{
+			Username:          username,
+			EncryptedPassword: account.password,
+			Nickname:          account.Nickname,
+			LastLoginAtUnix:   now,
+			Description:       account.Description,
+		}
+		err = updateUserCredentials(ctx, credentials, up)
+		if err != nil {
+			logrus.WithError(err).Warn("failed to update user credentials")
+		}
 	}
 
 	return &account.Account, nil
 }
 
+func ResetAccountPassword(ctx context.Context, username string, resetParam *param.ResetPasswordParam) (*acmodel.Account, error) {
+	credentials, err := getDashboardUserCredentials(ctx)
+	if err != nil {
+		return nil, err
+	}
+	account, err := fetchAccount(credentials, username)
+	if err != nil {
+		return nil, err
+	}
+	if account.LastLoginAt != nil && !account.LastLoginAt.IsZero() {
+		bts := sha256.Sum256([]byte(resetParam.OldPassword))
+		sha256EncodedPwd := hex.EncodeToString(bts[:])
+		if account.password != sha256EncodedPwd {
+			return nil, httpErr.NewBadRequest("password is incorrect")
```

Comment:
use "username or password is incorrect"

Reply:
The method is for reseting password after the user logs in. The username is gotten from session which wouldn't be wrong.

Reply:
OK